srv/modules/runners/rebuild.py

```diff
@@ -0,0 +1,257 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+Runner to remove and deploy OSDs
+"""
+
+from __future__ import absolute_import
+from __future__ import print_function
+import logging
+import pprint
+import time
+# pylint: disable=import-error,3rd-party-module-not-gated,redefined-builtin
+import salt.client
+import salt.runner
+
+log = logging.getLogger(__name__)
+
+
+def help_():
+    """
+    Usage
+    """
+    usage = ('salt-run rebuild.node [TARGET...]\n'
+             'salt-run rebuild.nodes [TARGET...]:\n\n'
+             '    Removes and deploys OSDs\n'
+             'salt-run rebuild.identical [TARGET...]:\n\n'
+             '    Removes and deploys OSDs with same IDs\n'
+             'salt-run rebuild.check [TARGET...]:\n\n'
+             '    Checks available space\n'
+             '\n\n'
+             'Examples:\n'
+             '    salt-run rebuild.node data1.ceph\n'
+             '    salt-run rebuild.node data1.ceph data2.ceph\n'
+             '    salt-run rebuild.node data[12].ceph\n'
+             '    salt-run rebuild.node I@roles:storage\n'
+             '    salt-run rebuild.node data*.ceph storage*.ceph\n'
+             '\n')
+    print(usage)
+    return ""
+
+
+def master_minion():
+    """
+    Load the master modules
+    """
+    __master_opts__ = salt.config.client_config("/etc/salt/master")
+    __master_utils__ = salt.loader.utils(__master_opts__)
+    __salt_master__ = salt.loader.minion_mods(
+        __master_opts__, utils=__master_utils__)
+
+    return __salt_master__["master.minion"]()
+
+
+# pylint: disable=too-many-instance-attributes
+class Rebuild(object):
+    """ Removes and deploys OSDs for a list of minions  """
+
+    def __init__(self, *args, **kwargs):
+        """ Sets clients, assigns minions """
+        self.master_minion = master_minion()
+        self.local = salt.client.LocalClient()
+        self.runner = salt.runner.RunnerClient(__opts__)
+        __master_opts__ = salt.config.client_config("/etc/salt/master")
+        __master_opts__['quiet'] = True
+        self.qrunner = salt.runner.RunnerClient(__master_opts__)
+        self.minions = self._minions(*args)
+        self.operation = kwargs.get('operation', 'osd.remove')
+        self.aggressive = kwargs.get('aggressive', False)
+        self.skipped = []
+
+    def _minions(self, *args):
+        """ Return expanded Salt target """
+        minions = set()
+        for arg in args:
+            results = self.local.cmd(arg, 'grains.get', ['id'], tgt_type="compound")
```

Comment:
learning from #1647 we might consider using grains.get('host') instead of `id`.

in `L:84` we are checking for `minion in osd_ret` where it can be beneficial to have the short hostname.

This needs to be confirmed in a virtualized environment

Reply:
The id vs host is not what is used.  The keys of the results are Salt minions.